lib/Monad_WP/OptionMonad.thy

```diff
@@ -60,32 +68,66 @@ lemma opt_map_upd_Some:
   "f(x \<mapsto> v) |> g = (f |> g)(x := g v)"
   by (auto simp: opt_map_def)
 
-lemmas opt_map_upd[simp] = opt_map_upd_None opt_map_upd_Some
+lemma opt_map_Some_upd_Some:
+  "f(x \<mapsto> v) ||> g = (f ||> g)(x \<mapsto> g v)"
+  by (simp add: opt_map_upd_Some)
+
+lemmas opt_map_upd[simp]
+  = opt_map_upd_None opt_map_upd_Some opt_map_Some_upd_Some
+
+lemma opt_map_upd_triv[simp]:
+  "t k = Some x \<Longrightarrow> (t |> f)(k := f x) = t |> f"
```

Comment:
This is potentially dangerous as a `[simp]` rule for performance, because it will always try to prove the precondition when the pattern occurs, but it is not that likely for that precondition to actually be true.

Might be better to collect this and the next two lemmas in a `lemmas` set, so that they are easy to add to a `simp` locally when they are needed.

Reply:
ah, ok, true. I'll remove from the simp set. This is quite useful for state updates etc but probably local enough.